### ISSUE TITLE
Enable auto-reloader,  program behavior not inconsistent with disable it.

### ISSUE THREAD

**Issue by zw963** [fxn/zeitwerk#29]
### folder directory:

├── app.rb
├── config.ru
├── Gemfile
└── Gemfile.lock

### content of config.ru

```rb
# config.ru
require 'roda'
require 'zeitwerk'

loader = Zeitwerk::Loader.new
# loader.logger = method(:puts)
loader.push_dir(__dir__)
loader.setup

if ENV['RACK_ENV'] == 'production'
  Zeitwerk::Loader.eager_load_all
  run App
else
  run ->(env) {
    loader.reload
    App.call(env)
  }
end
```

### content of  app.rb

```rb
class App < Roda
  articles = []

  route do |r|
    r.post "articles" do
      articles << r.params["content"]
      "Count: #{articles.count}"
    end
  end
end
```

### Expected behavior

When no enable zeitwerk, We can see Following output:

```
POST articles?content='content'
# => Count: 1
POST articles?content='content'
# => Count: 2
POST articles?content='content'
# => Count: 3
...
```
### Actual behavior

```
POST articles?content='content'
# => Count: 1
POST articles?content='content'
# => Count: 1
POST articles?content='content'
# => Count: 1
...
```

**Comment by fxn** [comment]
Since you application reloads the `App` class in every request, the variable `articles` is reset every time. You have to understand that as the `App` gets removed from memory, and the file `app.rb` evaluated again.

**Comment by zw963** [comment]
> Since you application reloads the `App` class in every request, the variable `articles` is reset every time. You have to understand that as the `App` gets removed from memory, and the file `app.rb` evaluated again.

Yes, i am understood now, maybe a better idea is to reload only file is changed.

Do you have plan to pull in [listen](https://github.com/guard/listen) gem to support this?

**Comment by zw963** [comment]
I do following change, program behavior is same now.

```diff
# config.ru
require 'roda'
require 'zeitwerk'
+ require 'listen'

loader = Zeitwerk::Loader.new
# loader.logger = method(:puts)
loader.push_dir(__dir__)
loader.setup
+Listen.to(__dir__) { loader.reload }.start

if ENV['RACK_ENV'] == 'production'
  Zeitwerk::Loader.eager_load_all
  run App
else
  run ->(env) {
-    loader.reload
    App.call(env)
  }
end
```

So, i thought if we can add listen into zeitwerk, more comfortable for gem user.
And, if rememer listen is a dependency for Rails, it should not pull extra dependency
to rails.

**Comment by zw963** [comment]
I release a dead simple gem wrap zeitwerk and listen, please check: 

https://github.com/zw963/hot_reloader

if you have any advice, please let me know, Thanks.

**Comment by fxn** [comment]
Hey!

Yeah, there I drawed a line. Zeitwerk won't provide a file watcher. The library tries to do one thing well: loading code.

Services need to implement a little coordination for thread-safety, and they may provide file watching. Client code is in a better position to know _how_ and _when_ to reload in their specific context.

Rails has integrated Zeitwerk, and it could be the case that someone writes a gem to integrate with Hanami, Roda, Sinatra, ..., or even that the frameworks themselves integrate, depending on the authors interests or design view. Gems that do not implement a running service do not need to reload, typically (when you change the code, you run the test suite which is going to pick whatever is in disk at the moment).

Closing this one, but thanks for the proposal and code snippets anyway.
